### PR TITLE
[Multipart] Add public types

### PR DESCRIPTION
**Pull request by czechboy0**
### Motivation

Add the public types approved in SOAR-0009.

### Modifications

Added the public types and make other changes outlined in SOAR-0009.

### Result

Most of the required runtime types are now in place.

### Test Plan

Added unit tests for the new code, where it made sense.

⚠️ The pipeline `pull request validation (api breakage)` is failing with:

```
1 breaking change detected in OpenAPIRuntime:
  💔 API breakage: constructor Configuration.init(dateTranscoder:) has removed default argument from parameter 0
** ERROR: ❌ Breaking API changes detected.
```

but that seems to be a false positive in the tool, as there is now a newer initializer that you can use as `.init()`, `.init(dateTranscoder:)`, `.init(multipartBoundaryGenerator:)`, or `init(dateTranscoder:multipartBoundaryGenerator:)`, so no existing code could be broken by this change.

### CODE REVIEW

File: Sources/OpenAPIRuntime/Base/CopyOnWriteBox.swift [hunk 1]
```diff
@@ -26,15 +26,15 @@
...
+        @usableFromInline init(value: Wrapped) { self.value = value }
     }
 
     /// The internal storage of the box.
     @usableFromInline internal var storage: Storage
 
     /// Creates a new box.
     /// - Parameter value: The value to store in the box.
-    @inlinable public init(value: Wrapped) { self.storage = .init(value: value) }
+    public init(value: Wrapped) { self.storage = .init(value: value) }
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Base/CopyOnWriteBox.swift:37]: Discovered by building the runtime library with library evolution enabled - you can't inline the designated initializer and still be resilient. The only alternative is to mark the struct as frozen.

File: Sources/OpenAPIRuntime/Interface/AsyncSequenceCommon.swift [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+// Copyright (c) 2023 Apple Inc. and the SwiftOpenAPIGenerator project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftOpenAPIGenerator project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
```
Comment (czechboy0) [inline Sources/OpenAPIRuntime/Interface/AsyncSequenceCommon.swift:14]: Factored out common code from `HTTPBody` to also be usable by `MultipartBody`.

File: Sources/OpenAPIRuntime/Interface/HTTPBody.swift [hunk 3]
```diff
@@ -122,24 +122,16 @@ public final class HTTPBody: @unchecked Sendable {
...
-        case multiple
-    }
-
-    /// The body's iteration behavior, which controls how many times
+    @available(
+        *,
+        deprecated,
+        renamed: "IterationBehavior",
+        message: "Use the top level IterationBehavior directly instead of HTTPBody.IterationBehavior."
+    ) public typealias IterationBehavior = OpenAPIRuntime.IterationBehavior
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Interface/HTTPBody.swift:130]: Can we move this to Deprecated.swift so it doesn't get missed in the 1.0.0 sweep?

File: Sources/OpenAPIRuntime/Multipart/MultipartBoundaryGenerator.swift [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        self.randomNumberSuffixLenght = randomNumberSuffixLenght
+    }
+    /// Generates a boundary string for a multipart message.
+    /// - Returns: A boundary string.
+    public func makeBoundary() -> String {
+        var randomSuffix = [UInt8](repeating: 0, count: randomNumberSuffixLenght)
+        for i in randomSuffix.startIndex..<randomSuffix.endIndex { randomSuffix[i] = values.randomElement()! }
+        return boundaryPrefix.appending(String(decoding: randomSuffix, as: UTF8.self))
+    }
+}
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Multipart/MultipartBoundaryGenerator.swift:75]: Some typos.

```suggestion
    /// The length, in bytes, of the random boundary suffix.
    public let randomNumberSuffixLength: Int

    /// The options for the random bytes suffix.
    private let values: [UInt8] = Array("0123456789".utf8)

    /// Create a new generator.
    /// - Parameters:
    ///   - boundaryPrefix: The constant prefix of each boundary.
    ///   - randomNumberSuffixLength: The length, in bytes, of the random boundary suffix.
    public init(boundaryPrefix: String = "__X_SWIFT_OPENAPI_", randomNumberSuffixLenght: Int = 20) {
        self.boundaryPrefix = boundaryPrefix
        self.randomNumberSuffixLength = randomNumberSuffixLength
    }
    /// Generates a boundary string for a multipart message.
    /// - Returns: A boundary string.
    public func makeBoundary() -> String {
        var randomSuffix = [UInt8](repeating: 0, count: randomNumberSuffixLength)
        for i in randomSuffix.startIndex..<randomSuffix.endIndex { randomSuffix[i] = values.randomElement()! }
        return boundaryPrefix.appending(String(decoding: randomSuffix, as: UTF8.self))
    }
}
```

File: Tests/OpenAPIRuntimeTests/Multipart/Test_MultipartBoundaryGenerator.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    func testConstant() throws {
+        let generator = ConstantMultipartBoundaryGenerator(boundary: "__abcd__")
+        let firstBoundary = generator.makeBoundary()
+        let secondBoundary = generator.makeBoundary()
+        XCTAssertEqual(firstBoundary, "__abcd__")
+        XCTAssertEqual(secondBoundary, "__abcd__")
+    }
+
+    func testRandom() throws {
+        let generator = RandomMultipartBoundaryGenerator(boundaryPrefix: "__abcd__", randomNumberSuffixLenght: 8)
```
Comment (simonjbeaumont) [inline Tests/OpenAPIRuntimeTests/Multipart/Test_MultipartBoundaryGenerator.swift:29]:
```suggestion
        let generator = RandomMultipartBoundaryGenerator(boundaryPrefix: "__abcd__", randomNumberSuffixLength: 8)
```

**Review by simonjbeaumont — CHANGES_REQUESTED** [review]
Minor changes, otherwise LGTM.

File: Sources/OpenAPIRuntime/Interface/HTTPBody.swift (hunk 3 above)
Comment (czechboy0) [reply Sources/OpenAPIRuntime/Interface/HTTPBody.swift:130]: Moved. FWIW, I always also search for the word `deprecated` in the project as some deprecations (e.g. stored properties) cannot be moved out of their primary location.

**Review by simonjbeaumont — APPROVED** [review]

**Comment by simonjbeaumont** [comment]
LGTM, feel free to land once you appease the CI (probably my bad suggestions causing formatting complaints; didn't check). I guess you'll need to merge over the API checker, which you said is a false positive.

**Pull request merged**